python/ql/src/experimental/Security/CWE-287-ConstantSecretKey/ConstantSecretKey.ql

```diff
@@ -0,0 +1,219 @@
+      )
+      or
+      exists(DataFlow::LocalSourceNode lsn |
+        lsn = API::moduleImport("os").getMember("environ").getASubscript().asSource() and
+        source.asExpr() = lsn.asExpr()
+      )
+    ) and
+    not source.getScope().getLocation().getFile().inStdlib()
+  }
+
+  /**
+   * Sinks are one of the following kinds, some of them are directly connected to a flask Instance like
+   * ```python
+   *    app.config['SECRET_KEY'] = 'CHANGEME1'
+   *    app.secret_key = 'CHANGEME2'
+   *    app.config.update(SECRET_KEY="CHANGEME3")
+   *    app.config.from_mapping(SECRET_KEY="CHANGEME4")
+   * ```
+   * other Sinks are SECRET_KEY Constants Variables that are defined in seperate files or a class in those files like:
+   * ```python
+   *    app.config.from_pyfile("config.py")
+   *    app.config.from_object('config.Config')
+   *```
+   * we find these files with `FromObjectFileName` DataFlow Configuration
+   * note that "JWT_SECRET_KEY" is same as "SECRET_KEY" but it is belong to popular flask-jwt-extended library
+   */
+  predicate isSink(DataFlow::Node sink) {
+    (
+      exists(API::Node n |
+        n = flaskInstance() and
+        flask_sessionSanitizer(n.getReturn().asSource())
+      |
+        sink =
+          [
+            n.getReturn().getAMember().getSubscript(["SECRET_KEY", "JWT_SECRET_KEY"]).asSink(),
+            n.getReturn().getMember(["SECRET_KEY", "JWT_SECRET_KEY"]).asSink(),
+            n.getReturn()
+                .getMember("config")
+                .getMember(["update", "from_mapping"])
+                .getACall()
+                .getArgByName(["SECRET_KEY", "JWT_SECRET_KEY"])
+          ]
+      )
+      or
+      // this query checks for Django SecretKey too
+      if exists(API::moduleImport("django"))
+      then
+        exists(AssignStmt e | e.getTarget(0).toString() = "SECRET_KEY" |
+          sink.asExpr() = e.getValue()
+          // and sanitizer(e.getTarget(0))
+        )
+      else
+        exists(SecretKeyAssignStmt e |
+          sink.asExpr() = e.getValue()
+          // | sanitizer(e.getTarget(0))
+        )
+    ) and
+    not sink.getScope().getLocation().getFile().inStdlib()
+  }
+}
+
+// using flask_session library is safe
+predicate flask_sessionSanitizer(DataFlow::Node source) {
+  not DataFlow::localFlow(source,
+    API::moduleImport("flask_session").getMember("Session").getACall().getArg(0))
+}
+
+// *it seems that sanitizer have a lot of performance issues*
+// for case check whether SECRECT_KEY is empty or not
+predicate sanitizer(Expr sourceExpr) {
+  exists(DataFlow::Node source, DataFlow::Node sink, If i |
+    source.asExpr() = sourceExpr and
+    DataFlow::localFlow(source, sink)
+  |
+    not i.getASubExpression().getAChildNode*().(Compare) = sink.asExpr() and
+    not sink.getScope().getLocation().getFile().inStdlib() and
+    not source.getScope().getLocation().getFile().inStdlib() and
+    not i.getScope().getLocation().getFile().inStdlib()
+  )
+}
```

Comment:
This predicate seems to select any expression that is not locally used in an if expression. If you only want to apply it for assignments of `SECRECT_KEY`, start by being able to select these in a predicate, and then ONLY look for which of _these_ can flow to an if statement.

However, configurations also have a predicate called `isSanitizer`, which you should be able to use for this (instead of restricting the sources).



As highlighted in the comment for `[app_safe_2.py](https://github.com/github/codeql/pull/13561/files#diff-0427b14b91512fc56afbdbe6b40f05ef99a40f497ae599c01e9ca95fb1d2d3c6)` I don't think the fact that we have an if statement makes this safe either.

Reply:
Since you are using the newest version of the API (great!) the predicate is called `isBarrier`.
A barrier is a node that does not allow flow to continue (in fact, it is removed from the flow graph). The comparison with a known safe value would be a prototypical example of a barrier. In fact, we have in our libraries [an implementation of "comparison with a constant"](https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/dataflow/new/BarrierGuards.qll#L6-L36) since that is a barrier in many queries.

You might be able to adapt the code by using just the first part, handling `eq` and `noteq` and replacing the string constant with the node for the default value.

In your case, you have a comparison with a known _unsafe_ value, so if you try to adapt the code to  you should flip the true and false branches.